src/utils/pcl.js

```diff
@@ -45,3 +50,157 @@ export const setOrUpdateGeometryAttribute = (geometry, attribute, data) => {
   }
   geometry.attributes[attribute].needsUpdate = true;
 };
+
+export class PCLDecoder {
+  static decode(message, colorChannel, useRainbow) {
+    const n = message.height * message.width;
+    const positions = new Float32Array(3 * n);
+    const colors = new Float32Array(3 * n);
+    const normals = new Float32Array(3 * n);
+
+    const uint8Buffer = Uint8Array.from(message.data).buffer;
+    const dataView = new DataView(uint8Buffer);
+    const offsets = {};
+    const accessor = {};
+
+    message.fields.forEach(f => {
+      offsets[f.name] = f.offset;
+      accessor[f.name] = getAccessorForDataType(dataView, message.datatype);
+    });
+
+    for (let i = 0; i < n; i++) {
+      const stride = i * message.point_step;
+      if (
+        offsets.x !== undefined &&
+        offsets.y !== undefined &&
+        offsets.z !== undefined
+      ) {
+        positions[3 * i] = accessor.x.call(
+          dataView,
+          stride + offsets.x,
+          !message.is_bigendian,
+        );
+        positions[3 * i + 1] = accessor.y.call(
+          dataView,
+          stride + offsets.y,
+          !message.is_bigendian,
+        );
+        positions[3 * i + 2] = accessor.z.call(
+          dataView,
+          stride + offsets.z,
+          !message.is_bigendian,
+        );
+      }
+
+      if (
+        offsets.rgb !== undefined &&
+        colorChannel === POINTCLOUD_COLOR_CHANNELS.RGB
+      ) {
+        colors[3 * i] = dataView.getUint8(stride + offsets.rgb + 2) / 255.0;
+        colors[3 * i + 1] = dataView.getUint8(stride + offsets.rgb + 1) / 255.0;
+        colors[3 * i + 2] = dataView.getUint8(stride + offsets.rgb) / 255.0;
+      }
+
+      if (
+        offsets.intensity !== undefined &&
+        colorChannel === POINTCLOUD_COLOR_CHANNELS.INTENSITY
+      ) {
+        const intensity = accessor.intensity.call(
+          dataView,
+          stride + offsets.intensity,
+          !message.is_bigendian,
+        );
+
+        const normalizedIntensity = Math.min(intensity, 360);
+
+        if (useRainbow) {
+          const rgb = convertColor(normalizedIntensity, 1.0, 0.5);
```

Comment:
Why not `THREE.Color().setHSL().toArray()`? The same threejs color instance can be reused too since we won't need to actuallly use it for anything other than the conversion